### ISSUE TITLE
Support for HTTP Basic Auth?

### ISSUE THREAD

**Issue by davidbarsky** [ashfurrow/Moya#151]
This is likely entirely my fault for missing it, but — I looked in the Kiosk app, and Artsy seems to encode the API keys into the the URL itself, but the API I’m working against encodes the keys using HTTP Basic Auth. Is that a supported feature of Moya, or is that something that I should drop down into Alamofire for (or even open a PR for)?

Thanks.

**Comment by davidbarsky** [comment]
(I should add — a lengthy answer isn’t necessary! just a yes or no would suffice.)

**Comment by ashfurrow** [comment]
tl;dr Moya doesn't currently support this.

Hmm. Interesting question. Alamofire seems to [handle this with `NSURLCredential`](https://github.com/Alamofire/Alamofire/blob/922c1ca3580b7adc9235bede1e99580e02ad3e9f/Source/Request.swift#L76-L80). Is this how you've done HTTP auth in the past? 

Maya doesn't currently have support for this, but using the `sharedSession` `NSURLSession`, so you _should_ be able to set things there and have all HTTP requests use those credentials. Obviously that's not a great fix, though. 

I guess what this would really need is some kind of (optional) way to specify which `NSURLSession` to use: the default, or one configured for your endpoints alone. 

Others are likely more familiar with the network library underneath Alamofire. Anyone have suggestions on how to implement this?

**Comment by davidbarsky** [comment]
Unless I misunderstand, HTTP Basic Auth is implemented by Alamofire as just a chained method after `.request`. 

Here’s my naive suggestion: pass the username/password to `MoyaProvider#sendRequest` as an optional parameter of type `NSURLCredential`, and if it’s not nil, call `.authenticate` after the `Alamofire.Manager.sharedInstance.request(request)` call.

Granted, that makes the method messy, hence my referring to it as “naive”.

**Comment by ashfurrow** [comment]
That makes sense. But it also adds a lot of complexity around individual API calls. Developers now need to remember which endpoints need auth and which don't, and that need is spread across their entire codebase. 

I think we need one of two things:

- an optional part of the `MoyaTarget` protocol that specifies per-target credentials
- an optional closure in the `MoyaProvider` initializer that gets last-moment chance to specify credentials based on the target. If it returns credentials, they're added to the request in the provider. 

I like the second option better because most users don't need HTTP basic auth and it keeps our protocol leaner. What do you think?

**Comment by davidbarsky** [comment]
Good point, I didn’t think through how my API would break a lot of applications in production. I like your suggestion of the the optional closure in the `MoyaProvider` initializer.

Would you like me to open a pull request in the next day or so?

**Comment by ashfurrow** [comment]
That would be lovely! Please let me know if you'd like to bounce ideas around or anything :smile:

**Comment by davidbarsky** [comment]
👍

**Comment by davidbarsky** [comment]
So the API I’m writing against is going to be deprecating their basic auth, and because of that, I don’t need to add that to Moya. Since nobody else seemed to expressed a need for it, I think it's best if the issue is closed.

**Comment by ashfurrow** [comment]
Cool, thanks for the update @davidbarsky. Please let us know if there's anything else you'd like to see added to Moya!

**Comment by davidbarsky** [comment]
Thanks for your help!
